lib/clients/PromAPIClient.js

```diff
@@ -0,0 +1,149 @@
+                    // build a regexp to select multiple labels
+                    selectors.push(`${labelType}=~"${selector.join('|')}"`);
+                } else {
+                    // select a simple label
+                    selectors.push(`${labelType}="${selector}"`);
+                }
+            });
+        }
+        if (labelRegexps) {
+            Object.keys(labelRegexps).forEach(labelType => {
+                const selector = labelRegexps[labelType];
+                selectors.push(`${labelType}=~"${selector}"`);
+            });
+        }
+        return `${metricName}{${selectors.join(',')}}`;
+    }
+
+    /**
+     * Send a PROMQL query to Prometheus API endpoint
+     *
+     * On success, get back the results as an array of values
+     * (extracted from the response JSON body "data.results"
+     * attribute).
+     *
+     * @param {object} params - query parameters
+     * @param {string} params.query - PROMQL query
+     * @param {werelogs.Logger} log - logger object
+     * @param {function} cb - callback function: cb(err, results)
+     * @return {undefined}
+     */
+    executeQuery(params, log, cb) {
+        const cbOnce = jsutil.once(cb);
+        const qs = querystring.stringify(params);
+        const opts = {
+            hostname: this._promHost,
+            port: this._promPort,
+            path: `/api/v1/query?${qs}`,
+            agent: this._promAPIHTTPAgent,
+        };
+        log.debug('sending query to Prometheus API server', {
+            method: 'LifecycleMetrics._executePromQuery',
+            host: opts.hostname,
+            port: opts.port,
+            params,
+        });
+        const begin = Date.now();
+        const req = http.request(opts, res => {
+            const buffers = [];
+            res.on('data', chunk => {
+                buffers.push(chunk);
+            });
+            return res.on('end', () => {
+                const data = buffers.join('');
+                log.debug(
+                    'response received for query to Prometheus API server', {
+                        method: 'LifecycleMetrics._executePromQuery',
+                        host: opts.hostname,
+                        port: opts.port,
+                        params,
+                        payloadLength: data.length,
+                        elapsedMs: Math.round(Date.now() - begin),
+                    });
+                let parsedResp;
+                try {
+                    parsedResp = JSON.parse(data);
+                } catch (err) {
+                    log.error(
+                        'error parsing response from Prometheus API server', {
+                            method: 'LifecycleMetrics._executePromQuery',
+                            host: opts.hostname,
+                            port: opts.port,
+                            path: opts.path,
+                            error: err.message,
+                            rawData: data.slice(0, 200),
```

Comment:
just curious why you chose to slice here to 200 chars?

Reply:
This is to avoid making a huge log line (which could be truncated itself by the logger, but not sure), since we are logging some raw return data, but it can be useful to log the first characters to get a better sense of the error (since it is not valid JSON).